### ISSUE TITLE
add an "other window'/"other frame" option for opening notes

### ISSUE THREAD

**Issue by JonatanSahar** [emacs-citar/citar#785]
**Is your feature request related to a problem? Please describe.**
I often want to review a note I have on a reference while writing text including that reference (by pressing `RET` on the cite link etc). atm, the note replaces the the buffer of the text I'm working on.

**Describe the solution you'd like**
I'd like to have the possibility of opening the note in another window/frame

**Describe alternatives you've considered**
I can always split the window first, and then access the cite link

Thanks!

**Comment by bdarcus** [comment]
I agree it makes sense, but am not really sure how to do that, or if it should really be the domain of this package. If you're in an org buffer, isn't that an org config issue?

Related:

https://emacs.stackexchange.com/questions/48683/open-org-mode-links-to-info-in-different-window

**Comment by swflint** [comment]
This sort of thing is related to the problem that we're having in joostkremers/ebib#301.  There, we're trying to use `citar--select-resource` and `(citar--get-notes-config :open)` to get a buffer to display, though it's unclear if we can expect a buffer from the notes opener.

**Comment by swflint** [comment]
Perhaps an `:open-no-display` could be helpful?

**Comment by joostkremers** [comment]
> I agree it makes sense, but am not really sure how to do that, or if it should really be the domain of this package. If you're in an org buffer, isn't that an org config issue?

I'm not a Citar user and I'm not familiar with its source code, or its design goals, so I can't definitively say, but in general, I'd say that it's the package that provides the interactive command that should also take care of displaying the buffer. But for that to be possible, it needs to be given a buffer, but it seems that that's not guaranteed to happen with Citar.

We're in the process of making Ebib's note-handling more open, i.e., making it easier to plug in a user's preferred note-taking system, and this is a problem we're running into. Ebib basically needs three things:

1. - Check if a BibTeX entry has a note.
2. - Display a note without selecting the buffer.
3. - Opening a note for editing / creating a new note, which involves displaying the note buffer and selecting it.

2. and 3. require that Ebib be given a buffer.

**Comment by Hugo-Heagren** [comment]
> in general, I'd say that it's the package that provides the interactive command that should also take care of displaying the buffer

I agree, in general, but the _general_ is doing some heavy lifting here. I think it's likely many users will want to use org-capture based workflows, and (for better or worse) there is no way to do so _without_ having the notes backend handle display (and not citar/ebib).

So I think that it would be good to design a system flexible enough to accomodate both:
- ebib/citar getting a buffer and handling display
- ebib/citar knowing that the notes backend has handled display

This goes along with my suggestion [here](https://github.com/joostkremers/ebib/pull/299#issuecomment-2098913333) ('Concerning 2') that the notes backend should return different values in different circumstances:
- a buffer return means that the driver app (ebib/citar) needs to handle displaying the buffer
- a special value (e.g. `t` or `nil` depending on the implementation) means that the backend has handled display

This system could easily be extended too. For example, if needed, a window return means the buffer is displayed in that window but the window is not selected (and the driver app has to do that). I'm not sure who would need this, but it's the sort of thing which would be easy to do if necessary. In the extreme case, this whole return-val/behaviour pattern could be encoded in a customisable user var (an alist mapping predicates on the return value to handler functions) so users could write their own handlers if need be. I think this is probably unnecessary, but it's the sort of thing you could do.

(Of course, it would *also* be really nice if the system allowed for different kinds of display, in other-window, other-tab, etc. but that's a separate thing.)

**Comment by joostkremers** [comment]
> So I think that it would be good to design a system flexible enough to accomodate both:
> 
> * ebib/citar getting a buffer and handling display
> * ebib/citar knowing that the notes backend has handled display

Yes, absolutely. It's what Ebib already allows when creating a note (because you can use org-capture for that). It would be easy to allow that for opening a note as well.

> * a buffer return means that the driver app (ebib/citar) needs to handle displaying the buffer
> * a special value (e.g. `t` or `nil` depending on the implementation) means that the backend has handled display

:+1: 

Ebib uses `nil`, BTW. Makes it easy to write something like:

```emacs-lisp
(if-let ((buffer (funcall ebib-notes-backend :open-note key)))
    (pop-to-buffer buffer))
```

> (Of course, it would _also_ be really nice if the system allowed for different kinds of display, in other-window, other-tab, etc. but that's a separate thing.)

That's what `display-buffer-alist` is for, isn't it? Whether it's the front-end or the back-end that handles displaying the notes buffer, they should use `display-buffer` and friends...

**Comment by bdarcus** [comment]
Seems much of the recent back-and-forth is about ebib. What, if anything, do we need here?

Does this impact our [note API](https://github.com/emacs-citar/citar/blob/main/citar.el#L438) at all?

**Comment by bdarcus** [comment]
> we're trying to use `citar--select-resource` and `(citar--get-notes-config :open)` to get a buffer to display, though it's unclear if we can expect a buffer from the notes opener.

They are both "private" function, so we didn't anticipate other packages using them.

@aikrahguzar @roshanshariff - any thoughts on this?

**Comment by joostkremers** [comment]
> Seems much of the recent back-and-forth is about ebib.

Indeed. :slightly_smiling_face: 

> What, if anything, do we need here?

Well, for Ebib it would be very helpful if there were guarantees about the return values of the functions in `citar-notes-sources`, esp. `:open` and `:create`. Judging by the doc string, those two don't require a specific return value.

> Does this impact our [note API](https://github.com/emacs-citar/citar/blob/main/citar.el#L438) at all?

Currently, Ebib assumes that `:open` returns a buffer and that Ebib should display it and can decide whether to select it o not; and that `:create` returns a buffer that Ebib should then display and select. `:create` can also return `nil`, in which case Ebib assumes that the back-end displays and selects the buffer.

The design on Ebib's side is not set in stone, so I'm open to alternatives.

**Comment by joostkremers** [comment]
> They are both "private" functions, so we didn't anticipate other packages using them.

But I assume Citar has a public API for opening and creating notes? (Or is that a silly assumption? Like I said, I don't use Citar.)

**Comment by bdarcus** [comment]
> But I assume Citar has a public API for opening and creating notes?

It does; it's a key feature.

Aside: I was noting that just to explain; not to suggest I'm apposed to making any changes.

Here's an example of its use in `citar-org-roam`.

https://github.com/emacs-citar/citar-org-roam/blob/82d47b5df1926627f56a09055c69b49b31cbbb9f/citar-org-roam.el#L75

So a possibility is we slightly change that API in ways that are more flexible, and work for your case.

**Comment by joostkremers** [comment]
> > But I assume Citar has a public API for opening and creating notes?
> 
> It does; it's a key feature.

:+1: 

> Aside: I was noting that just to explain; not to suggest I'm opposed to making any changes.

I didn't take it that way. :slightly_smiling_face: I just don't want to give you the impression I'm demanding changes to Citar.

> Here's an example of its use in `citar-org-roam`.
> 
> https://github.com/emacs-citar/citar-org-roam/blob/82d47b5df1926627f56a09055c69b49b31cbbb9f/citar-org-roam.el#L75
> 
> So a possibility is we slightly change that API in ways that are more flexible, and work for your case.

Our difficulty is that we don't want to define functions for the actions that Citar's API defines (`:hasitems`, `:open`, `:create`, etc.) because we assume that the user has already defined them. (Or has installed a package that defines them.) We would like a way to call the existing `:open` function and then either handle the buffer display and selection ourselves, or be confident that that's already been done.

**Comment by bdarcus** [comment]
> Our difficulty is that we don't want to define functions for the actions that Citar's API defines (`:hasitems`, `:open`, `:create`, etc.) because we assume that the user has already defined them. (Or has installed a package that defines them.) We would like a way to call the existing `:open` function and then either handle the buffer display and selection ourselves, or be confident that that's already been done.

Maybe we also add public variants of those functions that correspond to the updated, tighter, API requirement?

It seems a little odd, now that I look at it, that the default notes config for citar using those "private" functions.

https://github.com/emacs-citar/citar/blob/07d2a63c99fe35cbd468f8e6a322de05f1a29469/citar.el#L445

So the updated API description could be something like (feel to suggest improvements; this is just a strawman):

```
  :open function that, when given a note ID, returns a buffer ...

  :create function that, when given a note ID, creates new note if not present [not sure here]
```

**Comment by joostkremers** [comment]
> ```
>   :open function that, when given a note ID, returns a buffer ...
> 
>   :create function that, when given a note ID, creates new note if not present [not sure here]
> ```

This is what Ebib currently expects:

```
:open function that, when given a note ID, opens the note and returns the buffer containing it. The buffer should not be displayed nor selected. (The front-end does that, if necessary.) If opening the note fails, this function should silently fail, not raise an error.

:create function that, when given a note ID, creates a new note and returns the buffer without displaying or selecting it. Alternatively, it can display and select the buffer and return nil. If note creation fails, this function can raise an error.
```

If that sounds a little arbitrary, then you're probably right. Historic reasons... Especially the error handling is unfortunate, I think. So I'm not against making changes to Ebib.

Conceptually Ebib requires these actions (apart from `:has-note`):

```
:get-buffer like :open described above

:edit like :create above, but for existing notes 

:create as described above
```

Ebib currently collapses `:get-buffer` and `:edit` into `:open`, which is unfortunate, because `:edit` should be allowed to raise an error, and because the buffer should be displayed and selected (unlike for `:get-buffer`).

`:edit` and `:create` could be collapsed, because the distinction is just whether the note already exists, which the back-end knows. (The distinction between `:get-buffer ` and `:edit` lies in what the front-end wants to do with the note, which is why it's less practical to collapse them.)

Obviously, I'm reasoning from my perspective, which is based on what Ebib currently offers and what it would need in order to interface with other notes back-ends without changing any of its front-end code. That's obviously not a requirement, though.

**Comment by bdarcus** [comment]
@pprevos 

> Is there a version to test yet?

No. If and when there is, it will be linked here.

**Comment by swflint** [comment]
Is there anything I can do to help move this forward?

**Comment by roshanshariff** [comment]
Have you tried `other-window-prefix` (bound to `C-x 4 4` by default) before opening the note? Or `same-window-prefix` (`C-x 4 1`) or `other-frame-prefix` (`C-x 5 5`) if you prefer? In general, I think it would be good to be able to reuse as much as possible of Emacs' built-in window management configuration, in particular `display-buffer-alist` and friends. If there is missing functionality with that approach then we can consider what we can do in Citar.

**Comment by swflint** [comment]
I'm certainly missing some level of control to implement integration with other packages, see @joostkremers comments & joostkremers/ebib#299.

It's not just opening in other window or other frame that is missing, but these three options:

> :get-buffer     Like :open described above
> 
> :edit     Like :create above, but for existing notes 
> 
> :create     As described above

**Comment by roshanshariff** [comment]
@swflint, thanks, that helped me understand the situation better. I skimmed the discussion between you and @Hugo-Heagren in joostkremers/ebib#299 and I think you're getting at some of the issues.

Citar doesn't assume that notes correspond to filenames or buffers, and instead just handles them as opaque strings that are generated and consumed by the notes backend. This allows, for example, a Zotero backend that opens notes by making an inter-process call to the Zotero application. The backend only has to implement a minimal set of operations needed by Citar (i.e. listing notes for items, opening existing notes, and creating new notes).

I think there are good reasons to keep the backend operations minimal — @Hugo-Heagren also mentions an `org-capture` backend for which it would be difficult to satisfy additional requirements. Maybe for your purposes you can have a less general set of operations (buffer-oriented, say) and make an adapter that uses them to implement what Citar needs? Then notes backends satisfying your API could also be used for Citar via the adapter.

**Comment by Hugo-Heagren** [comment]
@roshanshariff I'm really glad there is more discussion on this, and I would be happy to add some thoughts/possibly code in the future, but I should say that if anyone is hoping for input from me I definitely don't have the capacity right now, and won't for the foreseeable future (sorry).

**Comment by swflint** [comment]
> Citar doesn't assume that notes correspond to filenames or buffers, and instead just handles them as opaque strings that are generated and consumed by the notes backend. This allows, for example, a Zotero backend that opens notes by making an inter-process call to the Zotero application. The backend only has to implement a minimal set of operations needed by Citar (i.e. listing notes for items, opening existing notes, and creating new notes).

I don't think "notes" should correspond to filenames or buffers, per se; they could very well be identifiers.  However, it does seem reasonable to assume that there is some way to get a buffer which contains the contents of note --- the buffer is the primary means of displaying and interacting with textual data in Emacs after all.  That is, in the Zotero example, being able to get the text from Zotero and display it in a buffer seems reasonable.  Whether that buffer is editable may be a different story.

> I think there are good reasons to keep the backend operations minimal — [@Hugo-Heagren](https://github.com/Hugo-Heagren) also mentions an `org-capture` backend for which it would be difficult to satisfy additional requirements. Maybe for your purposes you can have a less general set of operations (buffer-oriented, say) and make an adapter that uses them to implement what Citar needs? Then notes backends satisfying your API could also be used for Citar via the adapter.

I agree, the backend operations *should* be minimal.  But I'm not sure that the current set is quite right, at least as a consumer (programmer or user) of the Citar API.  This might be made most obvious by the request for other-window or other-frame; just one "open", and no way to wrap it as a user to control where it's opened.

**Comment by roshanshariff** [comment]
The way I see it, there are three ways to make the opening of notes more customizable:

1. Change the note backend API to retrieve buffers, then add customization variables in Citar, Ebib, or any other user of the API to control exactly what to do with those buffers.
2. Keep the note backend API the same, but for backends that open buffers in Emacs, add customization to control how exactly they open those buffers.
3. Option 2, but modify those backends to mark the buffer in a way that `display-action-alist` can control how those buffers are displayed.

If I understand correctly, you would like option 1. I feel like options 2 or 3 are more modular, in that Citar need only call a small set of verbs (open, create, and list) which match the commands it provides, and the note backend that handles the verb can control how exactly "open" is carried out, for example.

I think option 3 is even better: everything in emacs that opens buffers shouldn't need its own customization scheme to let the use control how that happens. Indeed, you can just invoke `display-buffer` which, in turn, consults `display-buffer-alist` as a universal customization point. For example, the built-in file note backend would do:

```elisp

;; Open a note using the `citar-file-note` category.
;; By default, open the note in the same window just like find-file.
(defun citar-file--open-note (filename)
  (display-buffer (find-file-noselect filename)
                  '(display-buffer-same-window
                    (category . citar-file-note))))
```

Then you, as a user, would customize how the buffer opens by
```elisp
;; Either reuse window already displaying the note, or open a new window at the bottom
(add-to-list 'display-buffer-alist
             '((category . citar-file-note)
               (display-buffer-reuse-window display-buffer-at-bottom)
               (post-command-select-window . t)))
```

**Comment by swflint** [comment]
> The way I see it, there are three ways to make the opening of notes more customizable:
> 
>     1. Change the note backend API to retrieve buffers, then add customization variables in Citar, Ebib, or any other user of the API to control exactly what to do with those buffers.
> 
>     2. Keep the note backend API the same, but for backends that open buffers in Emacs, add customization to control how exactly they open those buffers.
> 
>     3. Option 2, but modify those backends to mark the buffer in a way that `display-action-alist` can control how those buffers are displayed.
> 
> 
> If I understand correctly, you would like option 1. I feel like options 2 or 3 are more modular, in that Citar need only call a small set of verbs (open, create, and list) which match the commands it provides, and the note backend that handles the verb can control how exactly "open" is carried out, for example.

That isn't quite what I want.  I agree a small set of verbs is desirable, but I don't agree that the backend should be, in general, involved in display.  *Display* is a front-end action, getting something *to display* is its corresponding backend action.  Most front-end uses should likely be powered by `display-buffer`, absolutely!  But there should be a way to control it past that, or access the underlying content programmatically.  Hugo's suggestion of returning a buffer or sentinel indicating that the backend handles display seems reasonable.  Then the frontend can display or manipulate buffers as necessary.

> I think option 3 is even better: everything in emacs that opens buffers shouldn't need its own customization scheme to let the use control how that happens. Indeed, you can just invoke `display-buffer` which, in turn, consults `display-buffer-alist` as a universal customization point. For example, the built-in file note backend would do:
> 
> ;; Open a note using the `citar-file-note` category.
> ;; By default, open the note in the same window just like find-file.
> (defun citar-file--open-note (filename)
>   (display-buffer (find-file-noselect filename)
>                   '(display-buffer-same-window
>                     (category . citar-file-note))))
> 
> Then you, as a user, would customize how the buffer opens by
> 
> ;; Either reuse window already displaying the note, or open a new window at the bottom
> (add-to-list 'display-buffer-alist
>              '((category . citar-file-note)
>                (display-buffer-reuse-window display-buffer-at-bottom)
>                (post-command-select-window . t)))

As a note, it's not necessarily about just buffer opening.  Getting a buffer means I can manipulate it programmatically.

**Comment by roshanshariff** [comment]
The notes API in `citar.el` was designed to be a thin layer of indirection for the operations we provide. It's just a set of hooks delegating user commands like `citar-create-note` and `citar-open-note` to external providers like `citar-org-roam` and `citar-denote`. There is also some glue for exposing notes and other resources through a completing-read interface. The simple file-backed implementation in `citar-file.el` is meant as an example and because we already had most of the infrastructure for other purposes (and is nevertheless enough for many users). I believe the original bug is just asking for this implementation to be somewhat more flexible.

You can and should design an API at a more data-oriented level of abstraction. This API would belong in a package acting as a "source of truth" for notes, as Citar is for bibliographies. If you wanted a buffer for some note, you'd just ask that package instead of going through Citar. Then it would be pretty easy to delegate Citar's note commands to this package. It could even live in the Citar namespace, though it would be only loosely coupled with Citar's core functionality.

I do think that Citar should focus on bibliographies and references, not become a central, monolithic API provider for all bibliography-adjacent services. We also don't want to encroach on Ebib's role as a general-purpose bibliography manager (although over the years we've accumulated some functionality that belongs there).
